### DISCUSSION TITLE
BARE metal + FreeRTOS scratchpad linker issue

### DISCUSSION THREAD

**Discussion by mark-nagel** [polarfire-soc/polarfire-soc-documentation#30] [RTOS for PolarFire SoC] [answered]
Using the MPFS-lim.ld script the list shows .scratchpad starting at 0x0A00_0000 with length 0x0.
During execution of system_startup.c,:check_config_l2_scratchpad() an ASSERT occurs at line 177.  At this time the start and end registers both show a value of 0x0.
The linker script was modified to include __freertos_irq_stack_top in the stack section. 

Modification 1
 l2lim (rwx)      : ORIGIN = 0x08000000, LENGTH = 768k
HEAP_SIZE = 64k;

Modification 2
    /* must be on 4k boundary- corresponds to page size */
    .stack : ALIGN(0x1000)
    {​​​​​
        PROVIDE(__stack_bottom_h0$ = .);
        PROVIDE(__app_stack_bottom_h0 = .);
        . += STACK_SIZE_E51_APPLICATION;
        PROVIDE(__app_stack_top_h0 = .);
        PROVIDE(__stack_top_h0$ = .);

        PROVIDE(__freertos_irq_stack_top = .);

        PROVIDE(__stack_bottom_h1$ = .);
        PROVIDE(__app_stack_bottom_h1$ = .);
        . += STACK_SIZE_U54_1_APPLICATION;
        PROVIDE(__app_stack_top_h1 = .);
        PROVIDE(__stack_top_h1$ = .);
        ...

**Comment by hughbreslin** [comment]
Hey @mark-nagel 😄 

which freertos project are you running? Is it the one from the bare metal examples repository or is it from another source?

**Comment by mark-nagel** [comment]
Hey @hughbreslin 

I believe it's a more recent version.

**Reply by hughbreslin** [reply]
Could you send on a link so we can have a look? The one we had a look at is in [bare metal examples](https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/tree/main/driver-examples/mss/mss-ethernet-mac/mpfs-uart-mac-freertos_lwip) but we couldn't see what you were referring to

**Comment by mark-nagel** [comment]
Here's the link -- https://github.com/FreeRTOS/FreeRTOS/releases/download/202112.00/FreeRTOSv202112.00.zip
the '\FreeRTOS\DemoRISC-V_RV64_PolarFire_SoftConsole' example

**Comment by mark-nagel** [comment]
more info ...

Not using a FreeRTOS project.


  I tried to start with the mpfs-rpmsg project from the AMP repo - but it is a 'make file' project and I did not want to have to mess with the make files to port my code into it. 

  So, I started with the mss_rtc_time bare metal project - because it is an 'auto build' project.
     First I pulled in the FreeRTOS 9.x that is in the rpmsg project and after porting my code into it, I found that the 'xEventGroupSetBitsFromISR' command is supported in that build.

   So, I went to the FreeRTOS distro and pulled in the RISC-V_RV64_PolarFire_SoCoftnsole example and pulled that version of FreeRTOS in and added the 
ASM interrupt hook definition in the assembler definitions and the ISR re-direction in the .stack section of the LIM linker script. 

   I increased the L2LIM size to 768k and the heap to 64k in the linker script to accommodate my application.

**Reply by hughbreslin** [reply]
Hmm ok we're struggling to re-produce this on our end - would you be able to go back to the original copy of the RTC time project you have and build that and verify the start and end addresses are showing up correctly as 0x0a00_0000 and then make your modifications on top of that bit by bit so we can see where the issue is occurring? It doesn't show up with the out of the box version from the bare metal examples and we applied the changes you made to the linker script (less the PROVIDE(__freertos_irq_stack_top = .); ) which couldn't re-produce it. We increased the size of the stack by adding additional filler in place of the free rtos irq stack but that also doesn't reproduce this.

**Comment by hughbreslin** [comment, marked as answer]
Hi @mark-nagel 

Is the base LIM size big enough to hold your application? The default LIM size in the reference design is 512k but your estimated application size is  768k. The constants that are becoming zero are loaded into the stack which is at the end of memory and writes inside the LIM address space but to areas overflowing the LIM have no effect and reads from areas inside the LIM address space will return 0 if no memory is present at that address

**Comment by mark-nagel** [comment]
Hi @hughbreslin 

Good catch, that as the issue.  Using the Reference design as is, I hadn't checked the default memory configuration.  After increasing the LIM size in the MSS configurator to be larger than the application size, this issue was resolved.

Thanks!
